### ISSUE TITLE
InvalidOperationException in 13.0.0-rc4

### ISSUE THREAD

**Issue by Nuklon** [umbraco/Umbraco-CMS#15421]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

13.0.0-rc4

### Bug summary

It's happening when multiple pages are loaded closely together. To easily reproduce, add **two** subpages with the same doc type that are prefetched like this on a page:
``` 
2x:
        <link rel="prefetch" href="@prefetchUrl" as="document" />
```

```
System.InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
   at System.Collections.Generic.Dictionary`2.FindValue(TKey key)
   at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
   at Umbraco.Cms.Core.PropertyEditors.ValueConverters.BlockEditorPropertyValueConstructorCacheBase`1.TryGetValue(ValueTuple`2 key, Func`5& value)
   at Umbraco.Cms.Core.PropertyEditors.ValueConverters.BlockPropertyValueCreatorBase`4.BlockItemActivator`1.CreateInstance(Guid contentTypeKey, Nullable`1 settingsTypeKey, Udi contentUdi, IPublishedElement contentData, Udi settingsUdi, IPublishedElement settingsData)
   at Umbraco.Cms.Core.PropertyEditors.ValueConverters.BlockPropertyValueCreatorBase`4.<>c__DisplayClass12_0.<CreateBlockModel>g__CreateBlockItem|3(TBlockLayoutItem layoutItem)
   at System.Linq.Enumerable.SelectListIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.ToList()
   at Umbraco.Cms.Core.PropertyEditors.ValueConverters.BlockPropertyValueCreatorBase`4.CreateBlockModel(PropertyCacheLevel referenceCacheLevel, BlockEditorData converted, Boolean preview, IEnumerable`1 blockConfigurations, CreateEmptyBlockModel createEmptyModel, CreateBlockModelFromItems createModelFromItems, EnrichBlockItemModelFromConfiguration enrichBlockItem)
   at Umbraco.Cms.Core.PropertyEditors.ValueConverters.BlockPropertyValueCreatorBase`4.CreateBlockModel(PropertyCacheLevel referenceCacheLevel, String intermediateBlockModelValue, Boolean preview, IEnumerable`1 blockConfigurations, CreateEmptyBlockModel createEmptyModel, CreateBlockModelFromItems createModelFromItems, EnrichBlockItemModelFromConfiguration enrichBlockItem)
   at Umbraco.Cms.Core.PropertyEditors.ValueConverters.BlockListPropertyValueCreator.CreateBlockModel(PropertyCacheLevel referenceCacheLevel, String intermediateBlockModelValue, Boolean preview, BlockConfiguration[] blockConfigurations)
   at Umbraco.Cms.Core.PropertyEditors.ValueConverters.BlockListPropertyValueConverter.ConvertIntermediateToBlockListModel(IPublishedElement owner, IPublishedPropertyType propertyType, PropertyCacheLevel referenceCacheLevel, Object inter, Boolean preview)
   at Umbraco.Cms.Core.PropertyEditors.ValueConverters.BlockListPropertyValueConverter.ConvertIntermediateToObject(IPublishedElement owner, IPublishedPropertyType propertyType, PropertyCacheLevel referenceCacheLevel, Object inter, Boolean preview)
   at Umbraco.Cms.Core.Models.PublishedContent.PublishedPropertyType.ConvertInterToObject(IPublishedElement owner, PropertyCacheLevel referenceCacheLevel, Object inter, Boolean preview)
   at Umbraco.Cms.Infrastructure.PublishedCache.Property.GetValue(String culture, String segment)
   at Umbraco.Extensions.PublishedPropertyExtension.Value[T](IPublishedProperty property, IPublishedValueFallback publishedValueFallback, String culture, String segment, Fallback fallback, T defaultValue)
   at Umbraco.Extensions.PublishedContentExtensions.Value[T](IPublishedContent content, IPublishedValueFallback publishedValueFallback, String alias, String culture, String segment, Fallback fallback, T defaultValue)
```

### Specifics

Works in rc3.

### Steps to reproduce

See above.

**Comment by Zeegaan** [comment]
Can you be a little more specific in the reproduction steps, I just have set up a Root document with children. But I'm getting this error:
![image](https://github.com/umbraco/Umbraco-CMS/assets/70372949/f868cb9f-2350-4514-bc79-37d3f8ad30c6)


![image](https://github.com/umbraco/Umbraco-CMS/assets/70372949/f7cb6e13-3181-441f-8240-df213d1cce1b)

Child template:

```

@using Umbraco.Cms.Web.Common.PublishedModels;
@inherits Umbraco.Cms.Web.Common.Views.UmbracoViewPage<ContentModels.Child>
@using ContentModels = Umbraco.Cms.Web.Common.PublishedModels;
@{
	Layout = null;
}


<link rel="prefetch" href="@prefetchUrl" as="document" />
```

**Comment by Nuklon** [comment]
The prefetchUrl variable needs to be defined of course (to a subpage).
Say for example:
Home.cshtml (Home doc type) - /
Page.cshtml (Page doc type) - /pagea
Page.cshtml (Page doc type) - /pageb

In Home.cshtml render the prefetch URLs to /pagea and /pageb, e.g.:

```
<link rel="prefetch" href="/pagea" as="document" />
<link rel="prefetch" href="/pageb" as="document" />
```

The browser (I use Firefox) will prefetch these pages nearly simulateously when you open the home page (/). 
You can also reproduce it with something that loads /pagea and /pageb nearly simulateously, some simple .NET app will suffice.

Keep in mind that the Page doc type needs to have a property that uses the block edit thing. It's 100% reproducible here. It's also logical as looking at the code added here: https://github.com/umbraco/Umbraco-CMS/pull/15264
it's backed by a regular Dictionary, there's no locking or ConcurrentDictionary used, so when you have two pages requested at the same time this is bound to happen. My doc type has a lot of properties so it's probably quite slow, which might be the reason this shows up so easily for me.

**Comment by Nuklon** [comment]
Here: 
https://github.com/umbraco/Umbraco-CMS/pull/15264/files#diff-cbedd699e145689c90d8aac9fd3af61594b5f7d59d75396d9faf13fb9e580959R11

This should use a ConcurrentCache or lock around the Dictionary.

**Comment by kjac** [comment]
Hi @Nuklon,

Thank you so much for reporting this 💪 🚀 we should definitively use a ConcurrentDictionary here. 

On the positive side, this goes to prove that there is a win in having the constructors cached 🙈 

It _is_ a little tricky to reproduce, but I managed. For future reference, a sure shot way to reproduce is to add a custom value converter with a `Thread.Sleep()` in `ConvertIntermediateToObject` 😆 

I'll be submitting a PR to fix shortly.

**Comment by kjac** [comment]
Fixed in #15425
